library/src/com/google/maps/android/geoJsonLayer/GeoJsonRenderer.java

```diff
@@ -0,0 +1,308 @@
+import com.google.android.gms.maps.model.Polyline;
+import com.google.android.gms.maps.model.PolylineOptions;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Observable;
+import java.util.Observer;
+
+/**
+ * Renders GeoJsonFeature objects onto the GoogleMap as Marker, Polyline and Polygon objects. Also
+ * removes GeoJsonFeature objects and redraws features when updated.
+ */
+class GeoJsonRenderer implements Observer {
+
+    private final static int POLYGON_OUTER_COORDINATE_INDEX = 0;
+
+    private final static int POLYGON_INNER_COORDINATE_INDEX = 1;
+
+    private final static Object FEATURE_NOT_ON_MAP_VALUE = null;
+
+    private final HashMap<GeoJsonFeature, Object> mFeatures;
+
+    private GoogleMap mMap;
+
+    /**
+     * Creates a new GeoJsonRender object
+     *
+     * @param features hashmap of features to store
+     * @param map      map to place GeoJsonFeature objects on
+     */
+    public GeoJsonRenderer(HashMap<GeoJsonFeature, Object> features, GoogleMap map) {
+        mFeatures = features;
+        mMap = map;
+    }
+
+    /**
+     * Gets the GoogleMap that GeoJsonFeature objects are being placed on
+     *
+     * @return GoogleMap
+     */
+    public GoogleMap getMap() {
+        return mMap;
+    }
+
+    /**
+     * Changes the map that GeoJsonFeature objects are being drawn onto. Existing objects are
+     * removed from the previous map and drawn onto the new map.
+     *
+     * @param map GoogleMap to place GeoJsonFeature objects on
+     */
+    public void setMap(GoogleMap map) {
+        removeLayerFromMap();
+        if (map != null) {
+            // Re add to new map if it isn't null
+            mMap = map;
+            addLayerToMap();
+        }
+    }
+
+    /**
+     * Adds all GeoJsonFeature objects stored in the mFeatures hashmap onto the map
+     */
+    public void addLayerToMap() {
+        for (GeoJsonFeature geoJsonFeature : mFeatures.keySet()) {
+            addFeature(geoJsonFeature);
+        }
+    }
+
+    /**
+     * Adds a new GeoJsonFeature to the map if its geometry property is not null.
+     *
+     * @param feature feature to add to the map
+     */
+    public void addFeature(GeoJsonFeature feature) {
+        if (feature.hasGeometry()) {
+            mFeatures.put(feature, addFeatureToMap(feature, feature.getGeometry()));
+        }
+    }
+
+    /**
+     * Removes all GeoJsonFeature objects stored in the mFeatures hashmap from the map
+     */
+    public void removeLayerFromMap() {
+        for (GeoJsonFeature feature : mFeatures.keySet()) {
+            removeFeature(feature);
+        }
+    }
+
+    /**
+     * Removes a GeoJsonFeature from the map if its geometry property is not null
+     *
+     * @param feature feature to remove from map
+     */
+    public void removeFeature(GeoJsonFeature feature) {
+        if (feature.hasGeometry()) {
+            removeFromMap(mFeatures.get(feature));
```

Comment:
this method should also update the HashMap.  I actually think that you should do that in the removeFromMap method but you might have to change the logic slightly to be less recursive.